server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/CassandraSessionModule.java

```diff
@@ -91,19 +96,30 @@ Cluster provideCluster(CassandraSessionConfiguration cassandraSessionConfigurati
         List<Host> servers = listCassandraServers(configuration);
         QueryLoggerConfiguration queryLoggerConfiguration = getCassandraQueryLoggerConf(configuration);
 
-        return getRetryer(executor, configuration)
-                .getWithRetry(ctx -> ClusterWithKeyspaceCreatedFactory
-                        .config(
-                            ClusterBuilder.builder()
-                                .servers(servers)
-                                .queryLoggerConfiguration(queryLoggerConfiguration)
-                                .build(),
-                            configuration.getString("cassandra.keyspace"))
-                        .replicationFactor(configuration.getInt("cassandra.replication.factor"))
-                        .clusterWithInitializedKeyspace())
+        int maxRetries = configuration.getInt("cassandra.retryConnection.maxRetries", DEFAULT_CONNECTION_MAX_RETRIES);
+        int minDelay = configuration.getInt("cassandra.retryConnection.minDelay", DEFAULT_CONNECTION_MIN_DELAY);
+
+        return RetryExecutorUtil.retryOnExceptions(executor, maxRetries, minDelay, NoHostAvailableException.class)
+                .getWithRetry(getClusterRetryCallable(configuration, servers, queryLoggerConfiguration))
                 .get();
     }
 
+    private RetryCallable<Cluster> getClusterRetryCallable(PropertiesConfiguration configuration, List<Host> servers, QueryLoggerConfiguration queryLoggerConfiguration) {
+        LOGGER.info("Trying to connect to Cassandra service");
+
+        return ctx -> ClusterWithKeyspaceCreatedFactory
+            .config(getCluster(servers, queryLoggerConfiguration), configuration.getString("cassandra.keyspace"))
+            .replicationFactor(configuration.getInt("cassandra.replication.factor"))
+            .clusterWithInitializedKeyspace();
+    }
+
+    private Cluster getCluster(List<Host> servers, QueryLoggerConfiguration queryLoggerConfiguration) {
```

Comment:
Is just that I think you could directly use ClusterBuilder.build..... instead of using getCluster. I don't understand why you created a method that take two argument only to pass those two arguments to a builder.
But it's not a big deal, if you want you can keep it